### DISCUSSION TITLE
Transitioning to OpenGL

### DISCUSSION THREAD

**Discussion by Darylgolden** [ManimCommunity/manim#1400] [Architecture]
Our current system of having Mobjects for 2 different renderers is unsustainable. While a metaclass approach as suggested in #1272 could work, I think we should still try to move to making OpenGL the default renderer sooner rather than later.

Some of the challenges in transitioning could be:

- Docs (#1214)
- Tests (probably not difficult to do)
- Difficulties in helping users transition (#1240 is a known problem of the OpenGL renderer with Intel integrated graphics, we need to find a fix or workaround)
- Making sure the features/fixes/enhancements to Cairo are ported over to OpenGL (Update: Mobjects are being moved over using the metaclass approach)

**Comment by MrDiver** [comment]
# OpenGL rethinking the architecture (Maybe also deal with Cairo?)

This discussions seeks to find a proposable idea for a new rendering Architecture surrounding the OpenGL Renderer.

- [ ] As a first step i think we should find out what the current state of the renderer is in 3b1b/manim and make a sketch on how it currently works and what features are supported.
- [ ] Second step is to find out what we are lacking feature wise and how we might improve our own architecture or merge it with the state of 3b1b/manim
- [ ] Decide on what to do and file Issues corresponding to the proposed idea.


Related Issues and Ideas:
- #2454
- https://github.com/ManimCommunity/manim/pull/2635
- https://github.com/ManimCommunity/manim/pull/2455
- https://github.com/3b1b/manim/pull/1586
- https://github.com/3b1b/manim/issues/1331

**Comment by MrDiver** [comment]
# Rendering Curves with OpenGL

- [https://www.codeproject.com/articles/226569/drawing-polylines-by-tessellation](https://www.codeproject.com/articles/226569/drawing-polylines-by-tessellation "https://www.codeproject.com/articles/226569/drawing-polylines-by-tessellation")
- [Nvidia GPU Gem3 Vector Art Ch.25](https://developer.nvidia.com/gpugems/gpugems3/part-iv-image-effects/chapter-25-rendering-vector-art-gpu)
-  [Microsoft Rendering Bezier shapes](https://www.microsoft.com/en-us/research/wp-content/uploads/2005/01/p1000-loop.pdf "https://www.microsoft.com/en-us/research/wp-content/uploads/2005/01/p1000-loop.pdf")
-  [Curvy Blues](http://www.mdk.org.pl/2007/10/27/curvy-blues)
-  [OpenGL vs Cairo Beziers](http://www.mdk.org.pl/2007/8/6/vector-drawing-opengl-shaders-and-cairo)
-  [Offset Beziers](http://docs-hoffmann.de/bezier18122002.pdf)
-  [Funny Nvidia Forum Discussion](https://community.khronos.org/t/rendering-resolution-independent-curves/61875/11)
- [Vase Renderer](https://github.com/tyt2y3/vaserenderer "https://github.com/tyt2y3/vaserenderer")
- [OpenGL 4.0 Rasterization of Parametric curves using Tesselation](https://computeranimations.wordpress.com/2015/03/16/rasterization-of-parametric-curves-using-tessellation-shaders-in-glsl/ "https://computeranimations.wordpress.com/2015/03/16/rasterization-of-parametric-curves-using-tessellation-shaders-in-glsl/")
- [BlogPost Polyline](https://www.codeproject.com/articles/226569/drawing-polylines-by-tessellation)

- [Git List of Resources](https://github.com/openframeworks/openFrameworks/issues/1190)
- [GPU Path Rendering](https://developer.download.nvidia.com/devzone/devcenter/gamegraphics/files/opengl/gpupathrender.pdf)
- [Path Rendering can be done in GL Only?](https://www.youtube.com/watch?v=fTm_s_pCBEM&t=34s&ab_channel=sysrpn)
- [RA VG - Random Access Vector Graphics](https://www.hhoppe.com/ravg.pdf)
- [RAVG - Website](https://hhoppe.com/proj/ravg/)

The current best solution for rendering curves might be to implement nv_path_rendering in opengl only. Im not quite sure how much that would improve our situation but it might be worth a try.

Or doing something like VASEr which is kind of what we do now using earclip.

**Comment by MrDiver** [comment]
- #2455
Useful information
